hammers/scripts/lease_stack_reaper.py

```diff
@@ -147,22 +147,37 @@ def send_delete_notification(gpu_user, leases, sender):
         html)
 
 
-def collect_user_leases(db):
+def collect_user_leases(auth):
     users = {}
+    users_by_id = keystone.users(auth)
+    hosts_by_id = blazar.hosts(auth)
+    allocs_by_lease = collections.defaultdict(list)
+
+    for alloc in blazar.host_allocations(auth):
+        for r in alloc['reservations']:
+            allocs_by_lease[r['lease_id']].append(alloc['resource_id'])
+
 
-    for row in query.get_advanced_reservations_by_node_type(db):
-        user_id = row['user_id']
+    for lease_id, lease in blazar.leases(auth).items():
+        user_id = lease['user_id']
+        lease['nodes'] = allocs_by_lease[lease_id]
 
-        if row['project_id'] in EXCLUDED_PROJECT_IDS:
+        if lease['project_id'] in EXCLUDED_PROJECT_IDS:
             continue
 
-        if user_id not in list(users.keys()):
-            users[user_id] = User(
-                user_id=user_id,
-                name=row['user_name'],
-                email=json.loads(row['user_extra'])['email'])
+        if not users_by_id.get(user_id):
```

Comment:
Won't this throw if it's not in there? What about `if user_id not in users_by_id`

Reply:
We should probably also log when this happens, it seems like it would be unusual, and would imply that the user lookup failed. Imagine that it returns an empty dict/list for some reason. The lease reaper would quietly just ignore all of the leases and none would get reaped, and we'd be none the wiser.

Reply:
In testing, there was one lease that had it's user_id said the the user name. This was a quick fix. Turns out this was because I wasn't filtering terminated leases or only advanced reservations (user_ids were user_names in blazar in 2015-16).

Reply:
I've added filters.